src/main/kotlin/no/nav/familie/dokument/storage/StonadController.kt

```diff
@@ -19,7 +20,10 @@ import org.springframework.web.bind.annotation.*
 @ProtectedWithClaims(issuer = "selvbetjening", claimMap = ["acr=Level4"])
 class StonadController(@Autowired val storage: MellomLagerService,
                        @Autowired val contextHolder: TokenValidationContextHolder,
-                       @Autowired val objectMapper: ObjectMapper) {
+                       @Autowired val objectMapper: ObjectMapper,
+                       @Value("\${FAMILIE_DOKUMENT_FNR_SECRET_SALT}")
+                       val pepper : String
```

Comment:
blandning av salt og pepper. Burde man egentlige flytte allt av dette til ett sted i stedet for att controllers skal håndtere det? Eks i EncryptedStorage hvor vi allerede har kryptering av data?

Reply:
SHA256 - valgte denne da jeg tenkte det er "godt nok", bare en hash av et directory navn, ikke kryptering av innhold (passord), men kan godt prøve å bytte med noe annet. Ser andre på nav har valgt PBKDF2 (over BCrypt). Kanskje ikke bedre med NIST "approved" ;) 

Kan godt lage en impl med BCrypt :)

Hvis vi kunne stole på at vi har en unik id fra bruker i token som ikke endres over tid kunne vi kanskje brukt denne uten kryptering, men da har vi jo mindre kontroll selv. Hva skjer om noen bestemmer seg for å endre på hvordan denne id'en lages? 
  
Selve "hashingen" av fnr kan vi godt flytte til et sted vi gjør annen kryptering :)

Reply:
Når et salt er hemmelig kaller mange det pepper :) , men begrepet pepper er ikke brukt i specs. Der brukes begrepet secret-salt. Salt er ikke hemmelig.

Reply:
@SecurityChampions -> @geirmundsimonsen og @bent-lorentzen har dere noen tanker om dette? 

Kontekst: 
Vi mellomlagrer e kladd søknad for at bruker skal kunne fortsette på denne senere, eller laste den inn dersom bruker blir redirekted til inloggig etter en time. Vi lagrer også vedlegg her så disse kan lastes ned av api ved innsending av søknad. 

Tidligere lagret vi filer for en bruker i mapper/directories med fnr, dette for å finne igjen det det brukeren har mellomlagret. 

Autentisering og kryptering av innhold i søknad og vedlegg mm. er håndtert på andre måter. 
Men bruk av fnr i disse filene  åpner for å kunne liste ut alle fnr som har startet på søknad. For vedlegg hadde vi egentlig ikke trengt dette i det hele tatt (kunne bare brukt en uuid), men for den mellomlagrede søknaden trenger vi en måte å finne tilbake til data for bruker på en måte. 

Som Johan sier ovenfor kunne dette vært løst dersom token bruker autentiserer seg ed inneholdt en id på bruker som ikke endrer seg mellom hver innlogging (gjerne på forskjellige maskiner). 

Forslag over bruker ikke noe slikt fra token, men hasher fnr med pepper fra secret. Har dere noen tanker om dette?

Reply:
Etter litt fram og tilbake og en liten test med PBKDF2 tror jeg sha256 er bra nok (brukt sammen med pepper). PBKDF2 og BCrypt er laget for å kryptere passord og gir ikke noe ekstra sikkerhet for oss.Koden blir også enklere med sha256.

Vi har også begrenset megde data (vi sletter data etter noen dager), så sha256 burde være godt nok. Pepper gjør at ingen kan sjekke spesifikke fnr.

Reply:
Alternativ
fun lagPBKDF2Hash(fnr: String, salt: String): String {
    val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
    val spec = PBEKeySpec(fnr.toCharArray(), salt.toByteArray(), 10000, 256)
    val hash = factory.generateSecret(spec).encoded
    return Base64.getEncoder().encodeToString(hash)
}